zhaquirks/tuya/builder/__init__.py

```diff
@@ -0,0 +1,493 @@
+    ) -> QuirkBuilder:
+        """Add a Tuya Metering Configuration."""
+        self.tuya_dp(
+            dp_id,
+            metering_cfg.ep_attribute,
+            "current_summ_delivered",
+        )
+        self.adds(metering_cfg)
+        return self
+
+    def tuya_onoff(
+        self,
+        dp_id: int,
+        onoff_cfg: TuyaLocalCluster = TuyaOnOffNM,
+    ) -> QuirkBuilder:
+        """Add a Tuya OnOff Configuration."""
+        self.tuya_dp(
+            dp_id,
+            onoff_cfg.ep_attribute,
+            "on_off",
+        )
+        self.adds(onoff_cfg)
+        return self
+
+    def tuya_humidity(
+        self,
+        dp_id: int,
+        rh_cfg: TuyaLocalCluster = TuyaRelativeHumidity,
+        scale: float = 100,
+    ) -> QuirkBuilder:
+        """Add a Tuya Relative Humidity Configuration."""
+        self.tuya_dp(
+            dp_id,
+            rh_cfg.ep_attribute,
+            "measured_value",
+            converter=lambda x: x * scale,
+        )
+        self.adds(rh_cfg)
+        return self
+
+    def tuya_soil_moisture(
+        self,
+        dp_id: int,
+        soil_cfg: TuyaLocalCluster = TuyaSoilMoisture,
+        scale: float = 100,
+    ) -> QuirkBuilder:
+        """Add a Tuya Soil Moisture Configuration."""
+        self.tuya_dp(
+            dp_id,
+            soil_cfg.ep_attribute,
+            "measured_value",
+            converter=lambda x: x * scale,
+        )
+        self.adds(soil_cfg)
+        return self
+
+    def tuya_temperature(
+        self,
+        dp_id: int,
+        temp_cfg: TuyaLocalCluster = TuyaTemperatureMeasurement,
+        scale: float = 100,
+    ) -> QuirkBuilder:
+        """Add a Tuya Soil Moisture Configuration."""
+        self.tuya_dp(
+            dp_id,
+            temp_cfg.ep_attribute,
+            "measured_value",
+            converter=lambda x: x * scale,
+        )
+        self.adds(temp_cfg)
+        return self
+
+    def tuya_attribute(
+        self,
+        dp_id: int,
+        attribute_name: str,
+        type: type = t.uint16_t,
+        access: foundation.ZCLAttributeAccess = foundation.ZCLAttributeAccess.NONE,
+        is_manufacturer_specific=True,
+    ) -> QuirkBuilder:
+        """Add an attribute to AttributeDefs."""
+        attr_id: int = int.from_bytes([0xEF, dp_id])
+        self.new_attributes.update(
+            {
+                attr_id: foundation.ZCLAttributeDef(
+                    id=attr_id,
+                    type=type,
+                    access=access,
+                    is_manufacturer_specific=is_manufacturer_specific,
+                    name=attribute_name,
+                )
+            }
+        )
+
+        return self
+
+    def tuya_dp(
+        self,
+        dp_id: int,
+        ep_attribute: str,
+        attribute_name: str,
+        converter: Optional[Callable[[Any,], Any,]] = None,
+        dp_converter: Optional[Callable[[Any,], Any,]] = None,
```

Comment:
I'm aware formatting is disabled for this, but I think we can remove the trailing commas inside the typing at the end here, so it looks slightly nicer.
```suggestion
        converter: Optional[Callable[[Any], Any]] = None,
        dp_converter: Optional[Callable[[Any], Any]] = None,
```

Reply:
I think this still needs to be done.